packages/dds/tree/docs/SharedTree Philosophy.md

```diff
@@ -0,0 +1,254 @@
+
+## Data Model and Editing
+
+When users edit documents, those high level semantic operations need to be encoded in a way that supports merges.
+Additionally this encoding needs to be deterministically applied in all clients, even if they are using different version of SharedTree.
+This needs to be done such that applications can ensure invariants hold in their data model.
+This means the editing and merge logic must be kept identical for compatibility, and changes to them instead must only add support for new cases.
+For example, new edit operations can be added or new configuration flags added to existing ones, but how a particular merge gets resolved cannot be changed without introducing a compatibility issue.
+This can get complicated, so to keep it manageable, this problem was subdivided in a few different ways.
+
+The data-model for SharedTree was selected to assist with subdividing this problem.
+The selected data-model consists of many small structures named **fields** where collaborative edits interact, and **nodes** which connect them into a hierarchy and assigns types to each of the fields.
+
+The nodes have schema which define the schema for their fields.
+There are a few [kinds](<https://en.wikipedia.org/wiki/Kind_(type_theory)>) of nodes that configure their fields differently, for example objects (with a fixed set of mixed field types) or maps (with an extensible set of matching field types).
+
+There are also a few kinds of fields which provide different collaborative data-structures (like `sequence`, or `optional`), each of which hold nodes and define what edits can be made to them.
+
+All of the editing is done in the fields via the `field kind`.
+This allows the editing logic, both API and merge policy, to be packaged together into minimal versionable units (`field kinds`).
+Since the schema explicitly selects what field kinds to use, updating them is opt-in, and can be coordinated with deployment schedules to achieve the required cross version collaboration requirements.
+This also leverages all the same collaboration as the rest of schema evolution, making adopting alternative field kinds no different from updating other aspects of the application's schema.
+
+This results in a tree where the overall shape is controlled by the application via its tree schema (for the nodes) and the alternating layers of fields provide all the complex logic, neatly separating the concerns.
+This can also be thought of as a tree of entities alternating between fields and nodes.
+As these two kinds of entities alternate and thus never touch, this resembles a [bipartite graph](https://en.wikipedia.org/wiki/Bipartite_graph), and can be called a bipartite tree.
+
+To better align this data-model with existing libraries, `simple-tree` abstracts fields in a way that depends on the kind of the node they are part of.
```

Comment:
IMO, this feels a bit too "footnotey". I'd love to see a top-level section that covers the simple tree layer, as it is the result of the `Leverage existing tools, code and knowledge where practical` goal. While it is encoded in this text, I'd love the blurb to be more explicit in stating the "be JSON/POJO compatible" mantra. I'm not sure readers of this will come away thinking of this as one of the core principles.

Reply:
We are explicitly not POJO compatible since we support arrays and maps. And JSON is a text format not an API.

We are optionally partly compatible with code that reads enumerable own properties, with special support for ES6 maps and arrays. I'm not convinced that is part of our philosophy or how our goals impact the data model (which is the section we are in).

I don't want this to be too concreate/specific about what we are compatible with: this should instead be what we refer to when trying to decide/design what we are compatible with. If we decide to ditch es6 maps and moved to more record like APIs to be POJO compatible, for example, that should be because we decided that better aligns with the philosophy documented here, not a change we need to make in this document to keep it up to date.

If our philosophy was really centered around POJO, we would not have maps, arrays or customizable schema: I think thats too specific of a thing to be an implication of our goals on the data model. If we had a goal of being POJO compatible, that would be different.

Reply:
I'm not sure I fully agree with this, and I'm not sure if @DLehenbauer would either, but I'm fine with it merging in the current state and we can review it more later. I will restate that I do believe that POJO compat (and lossless JSON roundtripping) are eventual goals of ours that belong in this document. Regardless, thanks for the great doc!